### PR TITLE
[api] fix revert custom errors and returns the correct revert message…

### LINKED ISSUE #3919: [ Web3 API ] Custom Errors are not correctly returned
### What version of iotex-core image (or code branch) are you using?
N/A

### What operating system and processor architecture are you using?
Mac M1

### What did you do? If possible, provide a recipe for reproducing the error.
Unzip [hardhat-starter.zip](https://github.com/iotexproject/iotex-core/files/12306661/hardhat-starter.zip)

```
cd hardhat-starter
echo PRIVATE_KEY=[YOUR IOTEX DEV ACCOUNT] > .env
npm install
npx hardhat run scripts/deploy_and_reproduce.ts --network iotex_testnet
```

### What did you expect to see?
[Custom Errors](https://soliditylang.org/blog/2021/04/21/custom-errors/) have been introduced in Solidity 0.8.4. The script execution deploys a contract that reverts with a custom error and it should catch and log the custom error with its data.

### What did you see instead?
The script is failing with a Provider Error: _"ProviderError: execution simulation failed: status = 106"_

Run `npx hardhat run scripts/deploy_and_reproduce.ts` to see the correct behavior in the local Hardhat VM

### PR DESCRIPTION
**Pull request by millken**
# Description
Returns the correct json when the contract call contains a custom revert error

The PR needs to add a new HF height config in genesis.yaml , pls confirm it when deploying.

fix before
```bash
$ curl -X POST http://127.0.0.1:15014/ -H "Content-Type:application/json" \
            --data "{\"id\":2248897981,\"jsonrpc\":\"2.0\",\"method\":\"eth_estimateGas\",\"params\":[{\"from\":\"0x4f2f741648699c1dc0ad8352e937057cd7e66bd7\",\"to\":\"0x3f5f75147437c8354d1ff31ece231ba29be90f65\",\"data\":\"0xa6b206bf0000000000000000000000000000000000000000000000000000000000000000\",\"value\":\"0x0\",\"type\":\"0x1\"}]}"

{"jsonrpc":"2.0","id":2248897981,"error":{"code":13,"message":"execution simulation failed: status = 106"}}


$ curl -X POST http://127.0.0.1:15014/ -H "Content-Type:application/json" \
            --data '{"jsonrpc":"2.0","id":1,"method":"eth_call","params":[{"data":"0xa6b206bf0000000000000000000000000000000000000000000000000000000000000000","to":"0x3F5f75147437C8354d1fF31EcE231BA29bE90f65","gas":"0x2faf080"},"latest"]}'
{"jsonrpc":"2.0","id":1,"result":"0x1bd9aa400000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000007b000000000000000000000000000000000000000000000000000000000000001456616c75652063616e6e6f74206265207a65726f000000000000000000000000"}
```
after fix
```bash
$ curl -X POST http://127.0.0.1:15014/ -H "Content-Type:application/json" \
            --data "{\"id\":2248897981,\"jsonrpc\":\"2.0\",\"method\":\"eth_estimateGas\",\"params\":[{\"from\":\"0x4f2f741648699c1dc0ad8352e937057cd7e66bd7\",\"to\":\"0x3f5f75147437c8354d1ff31ece231ba29be90f65\",\"data\":\"0xa6b206bf0000000000000000000000000000000000000000000000000000000000000000\",\"value\":\"0x0\",\"type\":\"0x1\"}]}"

{"jsonrpc":"2.0","id":2248897981,"error":{"code":3,"message":"execution reverted","data":"0x1bd9aa400000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000007b000000000000000000000000000000000000000000000000000000000000001456616c75652063616e6e6f74206265207a65726f000000000000000000000000"}}

$ curl -X POST http://127.0.0.1:15014/ -H "Content-Type:application/json" \
            --data '{"jsonrpc":"2.0","id":1,"method":"eth_call","params":[{"data":"0xa6b206bf0000000000000000000000000000000000000000000000000000000000000000","to":"0x3F5f75147437C8354d1fF31EcE231BA29bE90f65","gas":"0x2faf080"},"latest"]}'

{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"execution reverted","data":"0x1bd9aa400000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000007b000000000000000000000000000000000000000000000000000000000000001456616c75652063616e6e6f74206265207a65726f000000000000000000000000"}}
```

The fixed API response is the same as geth

Fixes #3919

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockchain/genesis/genesis.go [hunk 1]
```diff
@@ -241,6 +243,8 @@ type (
 		// ToBeEnabledBlockHeight is a fake height that acts as a gating factor for WIP features
 		// upon next release, change IsToBeEnabled() to IsNextHeight() for features to be released
 		ToBeEnabledBlockHeight uint64 `yaml:"toBeEnabledHeight"`
+		// FixCustomErrorBlockHeight is the start height to fix contract's custom error in evm
+		FixCustomErrorBlockHeight uint64 `yaml:"fixCustomErrorHeight"`
```
Comment (dustinxie) [inline blockchain/genesis/genesis.go:247]: revert, use `ToBeEnabledBlockHeight` for new features

File: blockchain/genesis/genesis.go [hunk 2]
```diff
@@ -576,6 +580,11 @@ func (g *Blockchain) IsToBeEnabled(height uint64) bool {
 	return g.isPost(g.ToBeEnabledBlockHeight, height)
 }
 
+// IsCustomErrorEnabled checks whether height is equal to or larger than custom error height
+func (g *Blockchain) IsCustomErrorEnabled(height uint64) bool {
+	return g.isPost(g.FixCustomErrorBlockHeight, height)
+}
+
```
Comment (dustinxie) [inline blockchain/genesis/genesis.go:587]: revert

File: action/protocol/context.go [hunk 3]
```diff
@@ -224,6 +225,7 @@ func WithFeatureCtx(ctx context.Context) context.Context {
 			SystemWideActionGasLimit:                !g.IsAleutian(height),
 			NotFixTopicCopyBug:                      !g.IsAleutian(height),
 			SetRevertMessageToReceipt:               g.IsHawaii(height),
+			FixCustomErrorRevertMessage:             g.IsCustomErrorEnabled(height),
```
Comment (dustinxie) [inline action/protocol/context.go:228]: use `g.IsToBeEnabled()`

File: action/protocol/execution/evm/evm.go [hunk 4]
```diff
@@ -287,12 +288,21 @@ func ExecuteContract(
...
-		revertMsg := string(data[64 : 64+msgLength])
-		receipt.SetExecutionRevertMsg(revertMsg)
+	if featureCtx.SetRevertMessageToReceipt && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrExecutionReverted) && retval != nil {
+		if featureCtx.FixCustomErrorRevertMessage {
+			reason, errUnpack := abi.UnpackRevert(retval)
+			//we need to catch custom errors
+			if errUnpack != nil {
+				reason = "execution reverted"
+			}
+			receipt.SetExecutionRevertMsg(reason)
```
Comment (envestcc) [inline action/protocol/execution/evm/evm.go:298]: looks like similar code with `newRevertError`, why is `abi.UnpackRevert()` called twice

File: api/coreservice.go [hunk 5]
```diff
@@ -1430,19 +1430,19 @@ func (core *coreService) EstimateExecutionGasConsumption(ctx context.Context, sc
...
 	if err != nil {
 		return 0, status.Error(codes.Internal, err.Error())
 	}
 	if !enough {
-		if receipt.ExecutionRevertMsg() != "" {
-			return 0, status.Errorf(codes.Internal, fmt.Sprintf("execution simulation is reverted due to the reason: %s", receipt.ExecutionRevertMsg()))
+		if receipt != nil && len(receipt.ExecutionRevertMsg()) > 0 {
+			return 0, newRevertError(ret)
 		}
 		return 0, status.Error(codes.Internal, fmt.Sprintf("execution simulation failed: status = %d", receipt.Status))
```
Comment (envestcc) [inline api/coreservice.go:1441]: the return type of error should be consistent

File: action/protocol/execution/evm/evm.go (hunk 4 above)
Comment (millken) [reply action/protocol/execution/evm/evm.go:298]: here is only a check  whether it contains a custom error. Previously, an error was returned, but now it only catches the error without returning it. The specific error is converted into the  interface required by the JSON-RPC  according to the `newRevertError`

File: api/coreservice.go (hunk 5 above)
Comment (millken) [reply api/coreservice.go:1441]: `newRevertError`  In addition to being compatible with the error interface, it also provides the interface data required by JSON-RPC. code from go-ethereum

**Pull request closed without merging**